File: gomsic_core/parsers/quality_suite_log.py
"""Parser for ZEISS Quality Suite log files (log4j XML format).

Sources:
- ZQS/QualitySuite/user/Suite.log
- ZQS/QualitySuite/Administrator/Suite.log

These are log4j XML format with <log4j:event> elements containing
logger, level, timestamp, thread, message, exception, and properties.
"""

from __future__ import annotations

import logging
import re
from pathlib import Path
from typing import Optional

from ..debug_log.trace import ParserTraceContext
from ..extractor import ArchiveLayout
from ..models import QualitySuiteLogEntry, QualitySuiteLogSummary
from .base import BaseParser

logger = logging.getLogger(__name__)


class QualitySuiteLogParser(BaseParser):
    name = "quality_suite_log"

    def parse(self, layout: ArchiveLayout, ctx: ParserTraceContext) -> Optional[QualitySuiteLogSummary]:
        if not layout.zqs_dir:
            ctx.skip("ZQS directory not found")
            return None

        log_files: list[Path] = []
        qs_dir = layout.zqs_dir / "QualitySuite"
        if qs_dir.is_dir():
            for sub in ("user", "Administrator"):
                log_path = qs_dir / sub / "Suite.log"
                ctx.file_searched(str(log_path))
                if log_path.is_file():
                    ctx.file_found(str(log_path))
                    log_files.append(log_path)

        if not log_files:
            ctx.skip("No QualitySuite Suite.log files found")
            return None

        summary = QualitySuiteLogSummary()

        for log_file in log_files:
            text = self.read_text_file(log_file)
            if text is None:
                continue

            ctx.file_parsed(str(log_file))
            summary.files_analyzed.append(log_file.name)

            # Parse log4j:event elements
            event_pattern = re.compile(
                r'<log4j:event\s+logger="([^"]*?)"\s+level="([^"]*?)"\s+'
                r'timestamp="(\d+)"\s+thread="([^"]*?)">'
                r'(.*?)</log4j:event>',
                re.DOTALL,
            )

            for match in event_pattern.finditer(text):
                log_logger = match.group(1)
                level = match.group(2)
                timestamp = match.group(3)
                thread = match.group(4)
                body = match.group(5)

                # Extract message
                msg_match = re.search(r'<log4j:message>(.*?)</log4j:message>', body, re.DOTALL)
                message = msg_match.group(1).strip() if msg_match else ""

                # Extract exception/throwable
                exc_match = re.search(r'<log4j:throwable>(.*?)</log4j:throwable>', body, re.DOTALL)
                exception = exc_match.group(1).strip() if exc_match else None

                # Extract properties
                props: dict[str, str] = {}
                for pm in re.finditer(r'<log4j:data\s+name="([^"]*?)"\s+value="([^"]*?)"\s*/>', body):
                    props[pm.group(1)] = pm.group(2)

                # Only keep ERROR/WARN entries and key INFO entries
                if level in ("ERROR", "WARN", "FATAL"):
                    entry = QualitySuiteLogEntry(
                        timestamp=timestamp,
                        level=level,
                        logger=log_logger,
                        message=message,
                        exception=exception,
                        thread=thread,
                        properties=props,
                    )
                    summary.entries.append(entry)
                    if level in ("ERROR", "FATAL"):
                        summary.total_errors += 1
                    elif level == "WARN":
                        summary.total_warnings += 1

        ctx.note(f"QualitySuite logs: {summary.total_errors} errors, "
                 f"{summary.total_warnings} warnings from {len(summary.files_analyzed)} files")
        return summary if summary.entries else None

File: pwa/src/worker/parsers/quality-suite-log.ts
// ============================================================
// parsers/quality-suite-log.ts — Port of Python QualitySuiteLogParser
// Source: ZQS/QualitySuite/user/Suite.log (log4j XML events)
// ============================================================

import { BaseParser, ParserContext } from './base'
import type { QualitySuiteLogSummary, QualitySuiteLogEntry } from '../models'

const LEVELS_TO_KEEP = new Set(['ERROR', 'WARN', 'WARNING', 'FATAL'])

// log4j XML event pattern (dotall via /s flag)
const EVENT_RE =
  /<log4j:event\s+logger="([^"]*?)"\s+level="([^"]*?)"\s+timestamp="(\d+)"\s+thread="([^"]*?)">(.*?)<\/log4j:event>/gs
const MSG_RE = /<log4j:message>(.*?)<\/log4j:message>/s
const THROWABLE_RE = /<log4j:throwable>(.*?)<\/log4j:throwable>/s
const DATA_RE = /<log4j:data\s+name="([^"]*?)"\s+value="([^"]*?)"\s*\/>/g

function unescapeXml(s: string): string {
  return s
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
}

export class QualitySuiteLogParser extends BaseParser<QualitySuiteLogSummary> {
  readonly name = 'quality_suite_log'

  parse(ctx: ParserContext): QualitySuiteLogSummary | null {
    const { layout } = ctx

    const summary: QualitySuiteLogSummary = {
      entries: [],
      total_errors: 0,
      total_warnings: 0,
      files_analyzed: [],
    }

    // ZQS dir paths for Suite.log
    const zqsDir = layout.zqsDir
    if (!zqsDir) return null

    const candidatePaths = [
      `${zqsDir}/QualitySuite/user/Suite.log`,
      `${zqsDir}/QualitySuite/Administrator/Suite.log`,
    ]

    let foundAny = false

    for (const vpath of candidatePaths) {
      const text = this.getText(layout, vpath)
      if (!text) continue

      foundAny = true
      summary.files_analyzed.push(vpath)

      // Reset regex lastIndex before each file
      EVENT_RE.lastIndex = 0

      let em: RegExpExecArray | null
      while ((em = EVENT_RE.exec(text)) !== null) {
        const logger = em[1]
        const level = em[2].toUpperCase()
        const timestampMs = parseInt(em[3])
        const thread = em[4]
        const innerXml = em[5]

        if (!LEVELS_TO_KEEP.has(level)) continue

        // Extract message
        MSG_RE.lastIndex = 0
        const msgMatch = MSG_RE.exec(innerXml)
        const message = msgMatch ? unescapeXml(msgMatch[1].trim()) : ''

        // Extract throwable
        THROWABLE_RE.lastIndex = 0
        const throwMatch = THROWABLE_RE.exec(innerXml)
        const exception = throwMatch ? unescapeXml(throwMatch[1].trim()) : undefined

        // Extract properties
        const properties: Record<string, unknown> = {}
        DATA_RE.lastIndex = 0
        let dm: RegExpExecArray | null
        while ((dm = DATA_RE.exec(innerXml)) !== null) {
          properties[dm[1]] = unescapeXml(dm[2])
        }

        // Convert timestamp (milliseconds since epoch) to ISO string
        let timestamp: string | undefined
        if (!isNaN(timestampMs)) {
          try {
            timestamp = new Date(timestampMs).toISOString()
          } catch {
            timestamp = undefined
          }
        }

        const entry: QualitySuiteLogEntry = {
          timestamp,
          level,
          logger,
          message,
          exception,
          thread,
          properties,
        }

        summary.entries.push(entry)

        if (level === 'ERROR' || level === 'FATAL') {
          summary.total_errors++
        } else if (level === 'WARN' || level === 'WARNING') {
          summary.total_warnings++
        }
      }
    }

    if (!foundAny) return null
    return summary
  }
}
